pygpu/collectives.pyx

```diff
@@ -0,0 +1,299 @@
+from libc.stdlib cimport malloc, calloc, free
+
+from pygpu.gpuarray cimport (gpucontext, GpuContext, _GpuArray, GpuArray,
+                             ensure_context,
+                             GA_NO_ERROR, get_exc, gpucontext_error,
+                             GpuArray_IS_C_CONTIGUOUS,
+                             GA_C_ORDER, GA_F_ORDER, GA_ANY_ORDER,
+                             pygpu_empty_like, pygpu_empty, memcpy)
+from pygpu.gpuarray import GpuArrayException
+
+
+cdef class GpuCommCliqueId:
+    """
+    """
+    def __cinit__(self, GpuContext context=None):
+        self.context = ensure_context(context)
+
+    property comm_id:
+        "Unique clique id to be used by each GpuComm in a group of devices"
+        def __get__(self):
+            cdef bytes res
+            comm_generate_id(self.context.ctx, self)
```

Comment:
You generate a new clique_id each time the property is accessed.  I don't think that is efficient or wise.

You should probably generate it in the constructor and maybe have an optional parameter so that the user can feed in a bytes string to use as the id (like what you do in `__set__`).